### PR TITLE
Convex_hull_3: Switch to FaceGraph

### PR DESCRIPTION
**Pull request by afabri**
- The package now works with models of  `MutableFaceGraph` instead of [`ConvexHullPolyhedron_3`](http://doc.cgal.org/latest/Convex_hull_3/classConvexHullPolyhedron__3.html)
- I removed the function [`convex_hull_incremental_3()`](http://doc.cgal.org/latest/Convex_hull_3/group__PkgConvexHull3Functions.html#gac7a2053b0ddb855f79e0bfb22452332c)  which got deprecated with CGAL 4.6
- I deprecated the function [`convex_hull_3_to_polyhedron_3()`](http://doc.cgal.org/latest/Convex_hull_3/group__PkgConvexHull3Functions.html#gaa87070e950e63e3a7395a2fb14c7ea7c) and replaced it with `convex_hull_3_to_face_graph()` based on the new more general function `link_to_face_graph()`.   This function is in the package Triangulation_3.
- The class `Convex_hull_traits_3` got an additional template parameter for a `MutableFaceGraph` that defaults to  `CGAL::Polyhedron_3` , so that the convex hull function that writes into an `Object` becomes more general.

Small feature: https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Convex_hull_3_FaceGraph
> pre-approved by Sebastien 2016/12/08

### CODE REVIEW

**Comment by lrineau** [comment]
- That does not build on Linux
```
Building CXX object examples/Convex_hull_3/CMakeFiles/quickhull_3.dir/quickhull_3.cpp.o
In file included from /home/lrineau/Git/cgal/Convex_hull_3/examples/Convex_hull_3/../../include/CGAL/convex_hull_3.h:55:0,
                 from /home/lrineau/Git/cgal/Convex_hull_3/examples/Convex_hull_3/quickhull_3.cpp:3:
/home/lrineau/Git/cgal/Convex_hull_3/examples/Convex_hull_3/../../include/CGAL/convexity_check_3.h: In function ‘bool CGAL::is_strongly_convex_3(const Polyhedron&, const Traits&)’:
/home/lrineau/Git/cgal/Convex_hull_3/examples/Convex_hull_3/../../include/CGAL/convexity_check_3.h:79:11: error: need ‘typename’ before ‘boost::graph_traits<Graph>::face_descriptor’ because ‘boost::graph_traits<Graph>’ is a dependent scope
   typedef boost::graph_traits<Polyhedron>::face_descriptor face_descriptor;
           ^~~~~
/home/lrineau/Git/cgal/Convex_hull_3/examples/Convex_hull_3/../../include/CGAL/convexity_check_3.h:80:11: error: need ‘typename’ before ‘boost::graph_traits<Graph>::vertex_iterator’ because ‘boost::graph_traits<Graph>’ is a dependent scope
   typedef boost::graph_traits<Polyhedron>::vertex_iterator vertex_iterator;
           ^~~~~
/home/lrineau/Git/cgal/Convex_hull_3/examples/Convex_hull_3/../../include/CGAL/convexity_check_3.h:81:11: error: need ‘typename’ before ‘boost::graph_traits<Graph>::face_iterator’ because ‘boost::graph_traits<Graph>’ is a dependent scope
   typedef boost::graph_traits<Polyhedron>::face_iterator face_iterator;
           ^~~~~
```

- Given all those changes, including breaking changes, that will require some communication:
  - in a small feature page,
  - in `Installation/changes.html`.

**Comment by lrineau** [comment]
@afabri: I have a warning in my local Linux build:
```
Building CXX object test/Convex_hull_3/CMakeFiles/degeneracy_test.dir/degeneracy_test.cpp.o
In file included from /home/lrineau/Git/cgal/Convex_hull_3/test/Convex_hull_3/quickhull_test_3.cpp:5:0:
/home/lrineau/Git/cgal/Convex_hull_3/test/Convex_hull_3/../../include/CGAL/convex_hull_3.h: In function ‘void CGAL::internal::Convex_hull_3::coplanar_3_hull(InputIterator, InputIterator, const Point_3&, const Point_3&, const Point_3&, Polyhedron_3&, const Traits&)’:
/home/lrineau/Git/cgal/Convex_hull_3/test/Convex_hull_3/../../include/CGAL/convex_hull_3.h:287:50: warning: typedef ‘CH_2_iterator’ locally defined but not used [-Wunused-local-typedefs]
   typedef typename std::list<Point_3>::iterator  CH_2_iterator;
                                                  ^~~~~~~~~~~~~
```
If you fix it today, you should merge the branch in `integration`, to avoid that the warning is seen in the test results tomorrow.

File: Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h [hunk 1]
```diff
@@ -0,0 +1,873 @@
...
+{
+    typename Circ::value_type ed = static_cast<const Circ*>(this)->operator*();
+    e = E(ed);
+    return e;
+  }
+};
+  
+
+  //  The vertex iterator of the bgl must evaluate to a vertex handle, not to a vertex
+template < class T>
```
Comment (sloriot) [inline Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h]: What about using `Prevent_deref` instead?

File: Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h [hunk 2]
```diff
@@ -0,0 +1,873 @@
...
+
+  Self  operator--(int) {
+    Self tmp = *this;
+    --*this;
+    return tmp;
+  }
+};
+
+    
+    template <typename Tr>
```
Comment (sloriot) [inline Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h]: why do you need a class and can't directly use a pair?

File: Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h [hunk 3]
```diff
@@ -0,0 +1,873 @@
...
+    };
+
+
+  } // namespace detail
+} // namespace CGAL
+
+namespace boost { 
+
+  template <class GT, class TDS>
+  struct graph_traits< CGAL::Triangulation_data_structure_2<GT,TDS> > {
```
Comment (sloriot) [inline Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h]: Can't you inherit anything from the graph_traits of Triangulation_2?

File: Convex_hull_3/doc/Convex_hull_3/CGAL/Convex_hull_traits_3.h [hunk 4]
```diff
@@ -8,11 +8,12 @@ The class `Convex_hull_traits_3` serves as a traits class for the function
 function when `R` is a kernel with exact predicates but inexact constructions 
 (note that the type `Plane_3` is a triple of `Point_3` and not `R::Plane_3`). 
 
+\tparam MFG must be a ,model of  MutableFaceGraph
 \cgalModels `ConvexHullTraits_3`
 \cgalModels `IsStronglyConvexTraits_3` 
 
 */
-template< typename R >
+  template< typename R, typename MFG = Polyhedron_3<R> >
```
Comment (sloriot) [inline Convex_hull_3/doc/Convex_hull_3/CGAL/Convex_hull_traits_3.h]: Only looking at the diff on github, it seems that the function convex_hull_3(begin, end, P) has not been updated so that the polyhedron type is passed to the traits. `Default_traits_for_Chull_3` should have an extra template parameter to be able to pass the `FaceGraph` model to `Convex_hull_traits_3`.

File: Convex_hull_3/doc/Convex_hull_3/CGAL/Convex_hull_3/dual/halfspace_intersection_3.h [hunk 5]
```diff
@@ -15,7 +15,7 @@ This version does not construct the dual points explicitely but uses a special t
 \pre The computed intersection must be a bounded convex polyhedron.
 
 \tparam PlaneIterator must be an input iterator where the value type must be `Polyhedron::Traits::Plane_3`
-\tparam Polyhedron must be a model of `ConvexHullPolyhedron_3`.
```
Comment (sloriot) [inline Convex_hull_3/doc/Convex_hull_3/CGAL/Convex_hull_3/dual/halfspace_intersection_3.h]: `Convex_hull_3/test/Convex_hull_3/test_halfspace_intersections.cpp` should be updated to also test `Surface_mesh`.

File: Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h (hunk 1 above)
Comment (lrineau) [inline Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h]: Sébastien, be careful when you comment a diff. It seems the line you are commenting is the next one.

**Comment by lrineau** [comment]
Our scripts found out that `Convex_hull_3/examples/Convex_hull_3/data/cube.xyz` should not be executable.

File: Installation/changes.html [hunk 6]
```diff
@@ -148,6 +148,11 @@ <h2 id="release4.10">Release 4.10 </h2>
 <!-- Combinatorial Algorithms -->
 <!-- Geometry Kernels -->
 <!-- Convex Hull Algorithms -->
+  <h3>3D Convex Hull</h3>
+  <ul>
+    <li>The covex hull function can also produce a <code>Surface_mesh</code>,
```
Comment (sloriot) [inline Installation/changes.html]: covex -> convex

File: Convex_hull_3/doc/Convex_hull_3/CGAL/Convex_hull_traits_3.h [hunk 7]
```diff
@@ -42,7 +43,7 @@ typedef unspecified_type Plane_3;
 /*!
 
 */ 
-typedef Polyhedron_3<R> Polyhedron_3;
+typedef PolygonMesh Polygon_mesh;
```
Comment (lrineau) [inline Convex_hull_3/doc/Convex_hull_3/CGAL/Convex_hull_traits_3.h]: This is a breaking change.

It does even break the testsuite (seen in a local build of `integration`):
```
In file included from .../cgal/Point_set_processing_3/include/CGAL/vcm_estimate_normals.h:24:0,
                 from .../cgal/Point_set_processing_3/include/CGAL/vcm_estimate_edges.h:24,
                 from .../cgal/Point_set_processing_3/test/Point_set_processing_3/vcm_all_test.cpp:2:
.../cgal/Point_set_processing_3/include/CGAL/internal/Voronoi_covariance_3/voronoi_covariance_3.h: In instantiation of ‘F& CGAL::Voronoi_covariance_3::internal::tessellate_and_intersect(const DT&, typename DT::Vertex_handle, const Sphere&, F&) [with DT = CGAL::Delaunay_triangulation_3<CGAL::Epick, CGAL::Default, CGAL::Default, CGAL::Default>; Sphere = CGAL::Voronoi_covariance_3::Sphere_discretization<CGAL::Epick>; F = CGAL::Voronoi_covariance_3::internal::Covariance_accumulator_3<double>; typename DT::Vertex_handle = CGAL::internal::CC_iterator<CGAL::Compact_container<CGAL::Triangulation_vertex_base_3<CGAL::Epick, CGAL::Triangulation_ds_vertex_base_3<CGAL::Triangulation_data_structure_3<CGAL::Triangulation_vertex_base_3<CGAL::Epick, CGAL::Triangulation_ds_vertex_base_3<void> >, CGAL::Triangulation_cell_base_3<CGAL::Epick, CGAL::Triangulation_ds_cell_base_3<void> >, CGAL::Sequential_tag> > >, CGAL::Default, CGAL::Default, CGAL::Default>, false>]’:
.../cgal/Point_set_processing_3/include/CGAL/internal/Voronoi_covariance_3/voronoi_covariance_3.h:210:58:   required from ‘std::array<typename DT::Geom_traits::FT, 6ul> CGAL::Voronoi_covariance_3::voronoi_covariance_3(const DT&, typename DT::Vertex_handle, const Sphere&) [with DT = CGAL::Delaunay_triangulation_3<CGAL::Epick, CGAL::Default, CGAL::Default, CGAL::Default>; Sphere = CGAL::Voronoi_covariance_3::Sphere_discretization<CGAL::Epick>; typename DT::Geom_traits::FT = double; typename DT::Vertex_handle = CGAL::internal::CC_iterator<CGAL::Compact_container<CGAL::Triangulation_vertex_base_3<CGAL::Epick, CGAL::Triangulation_ds_vertex_base_3<CGAL::Triangulation_data_structure_3<CGAL::Triangulation_vertex_base_3<CGAL::Epick, CGAL::Triangulation_ds_vertex_base_3<void> >, CGAL::Triangulation_cell_base_3<CGAL::Epick, CGAL::Triangulation_ds_cell_base_3<void> >, CGAL::Sequential_tag> > >, CGAL::Default, CGAL::Default, CGAL::Default>, false>]’
.../cgal/Point_set_processing_3/include/CGAL/vcm_estimate_normals.h:92:53:   required from ‘void CGAL::internal::vcm_offset(ForwardIterator, ForwardIterator, PointPMap, std::vector<ValueT>&, double, std::size_t, const K&) [with ForwardIterator = __gnu_cxx::__normal_iterator<CGAL::Point_3<CGAL::Epick>*, std::vector<CGAL::Point_3<CGAL::Epick> > >; PointPMap = CGAL::Identity_property_map<CGAL::Point_3<CGAL::Epick> >; K = CGAL::Epick; Covariance = std::array<double, 6ul>; std::size_t = long unsigned int]’
.../cgal/Point_set_processing_3/include/CGAL/vcm_estimate_normals.h:256:26:   required from ‘void CGAL::compute_vcm(ForwardIterator, ForwardIterator, PointPMap, std::vector<std::array<typename Kernel::FT, 6ul> >&, double, double, const Kernel&) [with ForwardIterator = __gnu_cxx::__normal_iterator<CGAL::Point_3<CGAL::Epick>*, std::vector<CGAL::Point_3<CGAL::Epick> > >; PointPMap = CGAL::Identity_property_map<CGAL::Point_3<CGAL::Epick> >; Kernel = CGAL::Epick; typename Kernel::FT = double]’
.../cgal/Point_set_processing_3/test/Point_set_processing_3/vcm_all_test.cpp:47:78:   required from here
.../cgal/Point_set_processing_3/include/CGAL/internal/Voronoi_covariance_3/voronoi_covariance_3.h:141:59: error: no type named ‘Polyhedron_3’ in ‘class CGAL::Convex_hull_traits_3<CGAL::Epick, CGAL::Default, CGAL::Boolean_tag<false> >’
                     typedef typename Traits::Polyhedron_3 Polyhedron;
                                                           ^~~~~~~~~~
```

File: Convex_hull_3/include/CGAL/Convex_hull_traits_3.h [hunk 8]
```diff
@@ -185,20 +186,21 @@ struct Convex_hull_traits_base_3<R_, Tag_true>{
...
+  typedef Convex_hull_traits_3<R, Polyhedron, Has_filtered_predicates_tag>  Self;
   typedef typename R::Point_3                    Point_3;
   typedef typename R::Segment_3                  Segment_3;
   typedef typename R::Triangle_3                 Triangle_3;
   typedef Point_triple<R>                        Plane_3;
   typedef typename R::Vector_3                   Vector_3;
 
-  typedef CGAL::Polyhedron_3<R>                  Polyhedron_3;
+  typedef typename Default::Get<Polyhedron, CGAL::Polyhedron_3<R> >::type Polygon_mesh;
+  //  typedef Polyhedron_3;
```
Comment (lrineau) [inline Convex_hull_3/include/CGAL/Convex_hull_traits_3.h]: I suggest the typedef:
```C++
typedef Polygon_mesh Polyhedron_3;
```
for the compatibility.

File: Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h (hunk 1 above)
Comment (afabri) [inline Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h]: I guess `Prevent_deref` could be used all over the place, that is HDS, Arrangement, T2, Polyhedron.
Can we do that in another PR?

File: Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h [hunk 9]
```diff
@@ -4,20 +4,20 @@ namespace CGAL {
...
 \brief computes the convex hull of the set of points in the range
-[`first`, `last`). The polyhedron `P` is cleared, then
-the convex hull is stored in `P`. Note that the convex hull will be triangulated,
-that is `P` will contain only triangular facets.
+[`first`, `last`). The polyhedron `pm` is cleared, then
+the convex hull is stored in `pm`. Note that the convex hull will be triangulated,
+that is `pm` will contain only triangular facets.
  
-\attention This function does not compute the plane equations of the faces of `P`.
+\attention This function does not compute the plane equations of the faces of `pm`.
```
Comment (sloriot) [inline Convex_hull_3/doc/Convex_hull_3/CGAL/convex_hull_3.h]: This sentence is now useless

**Comment by lrineau** [comment]
The tests in CGAL-4.10-Ic-67 seemed good. Let's way for the period of two weeks after the pre-approval.

**Comment by lrineau** [comment]
The small feature is now accepted (after two weeks).

**Comment by lrineau** [comment]
Re-test tonight, because https://github.com/CGAL/cgal/pull/1701/commits/e24a7f6c3d6d1660190093374015f92e3be60b66 was not in CGAL-4.10-Ic-67.

File: Installation/changes.html [hunk 10]
```diff
@@ -148,6 +148,11 @@ <h2 id="release4.10">Release 4.10 </h2>
 <!-- Combinatorial Algorithms -->
 <!-- Geometry Kernels -->
 <!-- Convex Hull Algorithms -->
+  <h3>3D Convex Hull</h3>
+  <ul>
+    <li>The convex hull function can also produce a <code>Surface_mesh</code>,
+        and generally speaking any model of the concept <code>MutableFaceGraph</code></li>
+  </ul>
```
Comment (lrineau) [inline Installation/changes.html]: The PR introduces five important changes, but only one is described in the `changes.html`. The other changes are important, and should also be described here.

**Comment by sloriot** [comment]
@afabri Please answer/handle all the points of my review

File: Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h (hunk 1 above)
Comment (afabri) [inline Triangulation_2/include/CGAL/boost/graph/graph_traits_Triangulation_data_structure_2.h]: I just changed to Prevent_deref in this branch.

**Comment by lrineau** [comment]
@sloriot: unfortunately, your modification of `changes.html` conflicts with current version of `master`.
